connexion/middleware/security.py

```diff
@@ -0,0 +1,229 @@
+
+from connexion.apis.abstract import AbstractSpecAPI
+from connexion.exceptions import MissingMiddleware
+from connexion.http_facts import METHODS
+from connexion.lifecycle import MiddlewareRequest
+from connexion.middleware import AppMiddleware
+from connexion.security import SecurityHandlerFactory
+
+logger = logging.getLogger("connexion.middleware.security")
+
+
+class SecurityMiddleware(AppMiddleware):
+    """Middleware to check if operation is accessible on scope."""
+
+    def __init__(self, app: ASGIApp) -> None:
+        self.app = app
+        self.apis: t.Dict[str, SecurityAPI] = {}
+
+    def add_api(self, specification: t.Union[pathlib.Path, str, dict], **kwargs) -> None:
+        api = SecurityAPI(specification, **kwargs)
+        self.apis[api.base_path] = api
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send):
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        try:
+            connexion_context = scope['extensions']['connexion_routing']
+        except KeyError:
+            raise MissingMiddleware('Could not find operation_id in scope. Please make sure '
+                                    'you have a routing middleware registered upstream. ')
+
+        api_base_path = connexion_context.get('api_base_path')
+        if api_base_path:
+            api = self.apis[api_base_path]
+            operation_id = connexion_context.get('operation_id')
+            try:
+                operation = api.operations[operation_id]
+            except KeyError:
+                pass
+            else:
+                request = MiddlewareRequest(scope)
+                await operation(request)
+
+        await self.app(scope, receive, send)
+
+
+class SecurityAPI(AbstractSpecAPI):
+
+    def __init__(
+            self,
+            specification: t.Union[pathlib.Path, str, dict],
+            auth_all_paths: bool = False,
+            *args,
+            **kwargs
+    ):
+        super().__init__(specification, *args, **kwargs)
+        self.security_handler_factory = SecurityHandlerFactory('context')
+        self.app_security = self.specification.security
+        self.security_schemes = self.specification.security_definitions
+
+        if auth_all_paths:
+            self.add_auth_on_not_found()
+        else:
+            self.operations: t.Dict[str, SecurityOperation] = {}
+
+        self.add_paths()
+
+    def add_auth_on_not_found(self):
+        """Register a default SecurityOperation for routes that are not found."""
+        self.operations = defaultdict()
+        default_operation = self.make_operation()
+        self.operations.default_factory = lambda: default_operation
+
+    def add_paths(self):
+        paths = self.specification.get('paths', {})
+        for path, methods in paths.items():
+            for method, operation in methods.items():
+                if method not in METHODS:
+                    continue
```

Comment:
Wouldn't it make more sense to handle this more upstream when parsing the spec? If not, we should warn here that a method is being skipped.

Reply:
This is validated when parsing the spec, but there are other valid keys at this level. For instance for [common parameters](https://swagger.io/docs/specification/describing-parameters/) for a path. We want to skip those.